teps/0117-tekton-results-logs.md

```diff
@@ -0,0 +1,350 @@
+
+This TEP documents the design implemented in the
+[results#203 concept](https://github.com/tektoncd/results/pull/203).
+
+## Motivation
+
+Logs are critical pieces of information when analyzing or debugging Tekton
+`TaskRuns` and `PipelineRuns`. Today, logs are stored as standard Kubernetes
+container logs, which makes them subject to removal whenever the `Pod` that
+backs the `TaskRun` object is deleted. `Pods` and `TaskRuns` are often deleted
+("pruned") in production environments to free resources and storage in etcd.
+
+Many production environments utilize log forwarding and aggregation services to
+preserve `TaskRun` logs. This allows log data to be preserved off cluster and
+manage retention of logs externally. However, most of these services are
+designed to analyze logs in aggregate, which is not the typical use case for
+Tekton end users. Using log aggregation analyzers like Kibana to debug a failed
+`TaskRun` or `PipelineRun` can be quite cumbersome. There is a strong
+preference for a `kubectl logs` experience for externally stored Tekton logs.
+
+Furthermore, there are many use cases where log forwarding/aggregation is not
+suitable. Forwarding CI/CD logs to ElasticSearch is challenging at scale, and
+tools like Spyglass (used in Tekton's dogfooding CI cluster) only work with
+cloud provider object storage. Today, teams commonly add steps to their
+`PipelineRuns` which forward logs from preceding tasks to cloud provider object
+storage. Enhancing Results to perform this action would eliminate a significant
+amount of toil.
+
+### Goals
+
+- Extend the Tekton Results API to store and retrieve logs
+- Define how the Results APIServer can receive logs from the Watcher
+- Define a mechanism that allows multiple log backends to be supported
+
+### Non-Goals
+
+- Define an API/protocol which allows third-party log backends to be "plugged
+  in."
+- Support multiple log backends for a single Results instance.
+- Support log forwarding for custom tasks.
+
+### Use Cases
+
+- [Pipeline/Task User] Debug a failed `TaskRun` that had been deleted.
+- [Tekton Developer] Use Results to persist CI logs for Spyglass (dogfooding).
+- [Tekton Dashboard] Provide a fallback for accessing logs after the `TaskRun`
+  or backing `Pod` has been deleted.
+- [Platform Developer] Programmatically access `TaskRun` and `PipelineRun` logs
+  after the object is deleted from Kubernetes.
+
+
+### Requirements
+
+- Forwarding logs to Results must be opt-in.
+- Only one type of logging backend per Results instance will be supported. This
+  simplifies the implementation.
+
+## Proposal
+
+When the Results watcher finds a `TaskRun` object to preserve, it can be
+optionally configured to preserve the logs for the `TaskRun`. If the logs need
+to be preserved, the watcher will create a Record of type `TaskRunLog`, which
+indicates that logs have been persisted for a given TaskRun.
+
+When a record of type `TaskRunLog` is created, the APIServer will record the
```

Comment:
Given the second requirement of only supported backend for each results instance, do we need the api server to return the status.storage.category? If the TaskRunLog record creation succeeds, wouldn't that be enough confirmation for the watcher to start forwarding logs?

Reply:
Not all "backends" require the watcher to forward logs. See the updated not below on the `External` category.